[stable/sonarqube] overiding image.tag=7.4-community does not install default sonarqube plugins

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions. It's helpful to search the existing GitHub issues first. It's likely that another user has already reported the issue you're facing, or it's a known issue that we're already aware of-->

**Is this a request for help?**: No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG

<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as you can.  If you leave out
    information, we can't help you as well.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
-->

**Version of Helm and Kubernetes**:
helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

minikube 

**Which chart**: stable/sonarqube

**What happened**: 
The default version of sonarqube used on this chart is 6.7.6-community. 
.net core code needs 7.4-community to scan code correctly
Overriding the image.tag in values.yaml 7.4-community does use later docker image 
HOWEVER
when its running in the pod 7.4-community does not install the default plugins for sonarqube itself. 
even tho if you run the 7.4-community image via docker itself these plugins are installed correctly. 


**The reason for this is that....** 
- the 6.7.6-community image uses sonar zip from https://binaries.sonarsource.com/Distribution/sonarqube/sonarqube-6.7.6.zip which has a structure like: 
```
/sonarqube-6.7.6/extensions/plugins$ pwd
/sonarqube-6.7.6/extensions/plugins
/sonarqube-6.7.6/extensions/plugins$ ls
README.txt
/sonarqube-6.7.6/extensions/plugins$ 

```
_BUT it does have plugins in the lib/bundled-plugins/ folder_

```
/sonarqube-6.7.6/lib/bundled-plugins$ pwd
/sonarqube-6.7.6/lib/bundled-plugins
/sonarqube-6.7.6/lib/bundled-plugins$ ls
sonar-csharp-plugin-6.5.0.3766.jar  sonar-java-plugin-4.15.0.12310.jar      sonar-php-plugin-2.11.0.2485.jar    sonar-scm-git-plugin-1.3.0.869.jar  sonar-typescript-plugin-1.1.0.1079.jar
sonar-flex-plugin-2.3.jar           sonar-javascript-plugin-3.2.0.5506.jar  sonar-python-plugin-1.8.0.1496.jar  sonar-scm-svn-plugin-1.6.0.860.jar  sonar-xml-plugin-1.4.3.1027.jar
/sonarqube-6.7.6/lib/bundled-plugins$ 
```
- the 7.4-community image uses sonar zip from https://binaries.sonarsource.com/Distribution/sonarqube/sonarqube-7.4.zip which has a structure like: 

> /sonarqube-7.4/extensions/plugins$ pwd
> /sonarqube-7.4/extensions/plugins
> /sonarqube-7.4/extensions/plugins$ ls
> README.txt                          sonar-go-plugin-1.1.0.1612.jar          sonar-kotlin-plugin-1.2.1.2009.jar   sonar-ruby-plugin-1.2.1.2009.jar        sonar-vbnet-plugin-7.7.0.7192.jar
> sonar-csharp-plugin-7.7.0.7192.jar  sonar-jacoco-plugin-1.0.1.143.jar       sonar-ldap-plugin-2.2.0.608.jar      sonar-scm-git-plugin-1.6.0.1349.jar     sonar-xml-plugin-1.5.1.1452.jar
> sonar-css-plugin-1.0.2.611.jar      sonar-java-plugin-5.8.0.15699.jar       sonar-php-plugin-2.14.0.3569.jar     sonar-scm-svn-plugin-1.9.0.1295.jar
> sonar-flex-plugin-2.4.0.1222.jar    sonar-javascript-plugin-5.0.0.6962.jar  sonar-python-plugin-1.10.0.2131.jar  sonar-typescript-plugin-1.8.0.3332.jar
> /sonarqube-7.4/extensions/plugins$ 
and nothing in the lib/bundled-plugins/ folder

SonarQube has class 'ServerPluginRepository.java' (links to both 6 and 7 branches below)
- https://github.com/SonarSource/sonarqube/blob/branch-7.0/server/sonar-server/src/main/java/org/sonar/server/plugins/ServerPluginRepository.java
- https://github.com/SonarSource/sonarqube/blob/branch-6.7/server/sonar-server/src/main/java/org/sonar/server/plugins/ServerPluginRepository.java

In this class the function 'copyBundledPlugins' copies jars from lib/bundled-plugins/ to /extensions/plugins during start up. 
... hence on 6.7.6 the plugins are not in the /extensions/plugins to start with they are copied in later
and on 7.4 they are there to start with.

PROBLEM with the Helm Chart is that it includes a mounted volume to /opt/sonarqube/extensions
via the initContainer (used to download custom extra plugins) 
k8s mounted volume is overriding the folder on base 7.4-community image and therefore that folder is empty and no plugins installed to sonarqube 
https://stackoverflow.com/questions/53862726/stop-k8s-initcontainers-volume-overwriting-container-folder

**What you expected to happen**:
7.4-community to run on k8s cluster to have the base plugins installed on startup correctly. 

**How to reproduce it** (as minimally and precisely as possible):
- override the image.tag with 7.4-community 
- install chart and wait for sonarqube to startup 
- then on sonarqube menu go to 'Quality Profiles' you will see none are installed

**Anything else we need to know**: I have done work to correct this will submit a PR for this and add PR number when submitted.

When I would like to install plugins, it seems that the default plugins cannot be installed:

`Caused by: java.io.FileNotFoundException: /opt/sonarqube/extensions/plugins/sonar-csharp-plugin-6.5.0.3766.jar (Permission denied)`

I think this is related. Could you verify?

**chart version**

name: sonarqube
description: Sonarqube is an open sourced code quality scanning tool
version: 0.10.3

**image version**

```
image:
  repository: sonarqube
  tag: 6.7.6-community
```

**values**

```
plugins:
  install:
    - https://repox.sonarsource.com/sonarsource/org/sonarsource/go/sonar-go-plugin/1.1.0.1612/sonar-go-plugin-1.1.0.1612.jar
```

**logs**

Caused by: java.io.FileNotFoundException: /opt/sonarqube/extensions/plugins/sonar-csharp-plugin-6.5.0.3766.jar (Permission denied)

**Discussion**

Although the aim was to install the sonar-go-plugin, a permission denied related to a completely different plugin is shown. When the install-plugins element is commented out again, the error vanished.

--- a/stable/sonarqube/Chart.yaml
+++ b/stable/sonarqube/Chart.yaml
@@ -1,7 +1,7 @@
 name: sonarqube
 description: Sonarqube is an open sourced code quality scanning tool
-version: 0.11.0
-appVersion: 6.7.6
+version: 0.12.0
+appVersion: 7.4
 keywords:
   - coverage
   - security

--- a/stable/sonarqube/templates/copy-plugins.yaml
+++ b/stable/sonarqube/templates/copy-plugins.yaml
@@ -1,0 +1,16 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "sonarqube.fullname" . }}-copy-plugins
+  labels:
+    app: {{ template "sonarqube.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
+  copy_plugins.sh: |-
+      #! /bin/sh
+      cp /opt/sonarqube/extensions/plugins/tmp/*.jar /opt/sonarqube/extensions/plugins/
+      /opt/sonarqube/bin/run.sh
+    
+    

--- a/stable/sonarqube/templates/deployment.yaml
+++ b/stable/sonarqube/templates/deployment.yaml
@@ -26,16 +26,16 @@ spec:
             {{- end }}
           command: ["sh",
             "-c",
-            "mkdir -p /opt/sonarqube/extensions/plugins/ &&
-            cp /tmp/scripts/install_plugins.sh /opt/sonarqube/extensions/plugins/install_plugins.sh &&
-            chmod 0775 /opt/sonarqube/extensions/plugins/install_plugins.sh &&
-            /opt/sonarqube/extensions/plugins/install_plugins.sh
+            "mkdir -p /opt/sonarqube/extensions/plugins/tmp &&
+            cp /tmp/scripts/install_plugins.sh /opt/sonarqube/extensions/plugins/tmp/install_plugins.sh &&
+            chmod 0775 /opt/sonarqube/extensions/plugins/tmp/install_plugins.sh &&
+            /opt/sonarqube/extensions/plugins/tmp/install_plugins.sh && pwd && ls -lah
             "
            ]
           volumeMounts:
-            - mountPath: /opt/sonarqube/extensions
+            - mountPath: /opt/sonarqube/extensions/plugins/tmp
               name: sonarqube
-              subPath: extensions
+              subPath: tmp
             - name: install-plugins
               mountPath: /tmp/scripts/
           resources:
@@ -123,9 +123,11 @@ spec:
             - mountPath: /opt/sonarqube/data
               name: sonarqube
               subPath: data
-            - mountPath: /opt/sonarqube/extensions
+            - mountPath: /opt/sonarqube/extensions/plugins/tmp
               name: sonarqube
-              subPath: extensions
+              subPath: tmp
+            - name: copy-plugins
+              mountPath: /usr/local/
           resources:
 {{ toYaml .Values.resources | indent 12 }}
     {{- if .Values.nodeSelector }}
@@ -155,6 +157,13 @@ spec:
           items:
             - key: install_plugins.sh
               path: install_plugins.sh
+      - name: copy-plugins
+        configMap:
+          name: {{ template "sonarqube.fullname" . }}-copy-plugins
+          defaultMode: 0755
+          items:
+            - key: copy_plugins.sh
+              path: copy_plugins.sh                
       - name: startup
         configMap:
           name: {{ template "sonarqube.fullname" . }}-startup

--- a/stable/sonarqube/templates/install-plugins.yaml
+++ b/stable/sonarqube/templates/install-plugins.yaml
@@ -9,7 +9,7 @@ metadata:
     heritage: {{ .Release.Service }}
 data:
   install_plugins.sh: |-
-    cd /opt/sonarqube/extensions/plugins
+    cd /opt/sonarqube/extensions/plugins/tmp
     {{- if .Values.plugins.install }}
       {{- range $index, $val := .Values.plugins.install }}
         wget {{ $val }} --no-check-certificate

--- a/stable/sonarqube/values.yaml
+++ b/stable/sonarqube/values.yaml
@@ -4,7 +4,7 @@
 replicaCount: 1
 image:
   repository: sonarqube
-  tag: 6.7.6-community
+  tag: 7.4-community
   # If using a private repository, the name of the imagePullSecret to use
   # pullSecret: my-repo-secret
 # Starting 6.7.6 and 7.4 official sonarqube docker image, command is not required

